pkg/controllers/monitoring-stack/controller.go

```diff
@@ -164,7 +166,41 @@ func (r *reconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Resu
 
 		}
 	}
+	return r.cleanupResources(ctx, ms)
+}
+
+func (r *reconciler) cleanupResources(ctx context.Context, ms *stack.MonitoringStack) (ctrl.Result, error) {
+	if ms.ObjectMeta.DeletionTimestamp.IsZero() {
+		if !containsString(ms.ObjectMeta.Finalizers, finalizerName) {
+			controllerutil.AddFinalizer(ms, finalizerName)
+			if err := r.k8sClient.Update(ctx, ms); err != nil {
+				if errors.IsConflict(err) {
+					return ctrl.Result{RequeueAfter: 2 * time.Second}, nil
+				}
+				return ctrl.Result{}, err
+			}
+		}
+	} else if containsString(ms.ObjectMeta.Finalizers, finalizerName) {
```

Comment:
should we check if the finalizer is set before deleting grafanadatasource?

Reply:
It would have been set in the `if` condition, this is just to confirm.